File: app/src/main/java/org/lineageos/jelly/history/HistoryAnimationDecorator.java
/*
 * Copyright (C) 2017 The LineageOS Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.lineageos.jelly.history;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;
import android.view.View;

import org.lineageos.jelly.R;

class HistoryAnimationDecorator extends RecyclerView.ItemDecoration {

    private final Drawable mBackground;

    HistoryAnimationDecorator(Context context) {
        mBackground = new ColorDrawable(ContextCompat.getColor(context, R.color.colorDelete));
    }

    @Override
    public void onDraw(Canvas c, RecyclerView parent, RecyclerView.State state) {
        if (!parent.getItemAnimator().isRunning()) {
            super.onDraw(c, parent, state);
            return;
        }

        View firstComingUp = null;
        View lastComingDown = null;
        int left = 0;
        int top = 0;
        int right = parent.getWidth();
        int bottom = 0;

        int size = parent.getLayoutManager().getChildCount();
        for (int i = 0; i < size; i++) {
            View child = parent.getLayoutManager().getChildAt(i);
            if (child.getTranslationY() < 0) {
                lastComingDown = child;
            } else if (child.getTranslationY() > 0 && firstComingUp == null) {
                firstComingUp = child;
            }
        }

        if (firstComingUp != null && lastComingDown != null) {
            top = lastComingDown.getBottom() + (int) lastComingDown.getTranslationY();
            bottom = firstComingUp.getTop() + (int) firstComingUp.getTranslationY();
        } else if (firstComingUp != null) {
            top = firstComingUp.getTop();
            bottom = firstComingUp.getTop() + (int) firstComingUp.getTranslationY();
        } else if (lastComingDown != null) {
            top = lastComingDown.getBottom() + (int) lastComingDown.getTranslationY();
            bottom = lastComingDown.getBottom();
        }

        mBackground.setBounds(left, top, right, bottom);
        mBackground.draw(c);

        super.onDraw(c, parent, state);
    }


}

File: app/src/main/java/org/lineageos/jelly/history/HistoryAnimationDecorator.kt
/*
 * SPDX-FileCopyrightText: 2020 The LineageOS Project
 * SPDX-License-Identifier: Apache-2.0
 */

package org.lineageos.jelly.history

import android.content.Context
import android.graphics.Canvas
import android.graphics.drawable.ColorDrawable
import android.view.View
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ItemDecoration
import org.lineageos.jelly.R

internal class HistoryAnimationDecorator(context: Context) : ItemDecoration() {
    private val background = ColorDrawable(ContextCompat.getColor(context, R.color.colorDelete))

    override fun onDraw(c: Canvas, parent: RecyclerView, state: RecyclerView.State) {
        if (!parent.itemAnimator!!.isRunning) {
            super.onDraw(c, parent, state)
            return
        }
        var firstComingUp: View? = null
        var lastComingDown: View? = null
        val left = 0
        var top = 0
        val right = parent.width
        var bottom = 0
        val layoutManager = parent.layoutManager!!
        val size = layoutManager.childCount
        for (i in 0 until size) {
            val child = layoutManager.getChildAt(i)!!
            if (child.translationY < 0) {
                lastComingDown = child
            } else if (child.translationY > 0 && firstComingUp == null) {
                firstComingUp = child
            }
        }
        if (firstComingUp != null && lastComingDown != null) {
            top = lastComingDown.bottom + lastComingDown.translationY.toInt()
            bottom = firstComingUp.top + firstComingUp.translationY.toInt()
        } else if (firstComingUp != null) {
            top = firstComingUp.top
            bottom = firstComingUp.top + firstComingUp.translationY.toInt()
        } else if (lastComingDown != null) {
            top = lastComingDown.bottom + lastComingDown.translationY.toInt()
            bottom = lastComingDown.bottom
        }
        background.setBounds(left, top, right, bottom)
        background.draw(c)
        super.onDraw(c, parent, state)
    }
}
